drivers/flash/flash_rpi_pico.c

```diff
@@ -0,0 +1,189 @@
+/*
+ * Copyright (c) 2022 Yonatan Schachter
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <errno.h>
+#include <zephyr/device.h>
+#include <zephyr/drivers/flash.h>
+#include <zephyr/logging/log.h>
+#include <hardware/flash.h>
+
+LOG_MODULE_REGISTER(flash_rpi, CONFIG_FLASH_LOG_LEVEL);
+
+#define DT_DRV_COMPAT raspberrypi_pico_flash_controller
+
+#define PAGE_SIZE 256
+#define ERASE_VALUE 0xff
+#define FLASH_PAGE_MASK (0xffffffff - (PAGE_SIZE - 1))
+#define FLASH_SIZE (CONFIG_FLASH_SIZE * 1024)
+
+static uint8_t page_buffer[PAGE_SIZE];
+
+static const struct flash_parameters flash_rpi_parameters = {
+	.write_block_size = 1,
+	.erase_value = ERASE_VALUE,
+};
+
+static bool is_invalid_range(off_t offset, uint32_t size)
+{
+	return offset + size > FLASH_SIZE;
+}
+
+static int flash_rpi_read(const struct device *dev, off_t offset,
+				void *data, size_t size)
+{
+	if (is_invalid_range(offset, size)) {
+		return -EINVAL;
+	}
+
+	if (!size) {
+		return 0;
+	}
+
+	memcpy(data, (uint8_t *)(CONFIG_FLASH_BASE_ADDRESS + offset), size);
+
+	return 0;
+}
+
+static int flash_rpi_write(const struct device *dev, off_t offset,
+				const void *data, size_t size)
+{
+	size_t left_pad_len = offset & ~FLASH_PAGE_MASK;
+	size_t write_count;
+	uint8_t *_data = (uint8_t *)data;
+
+	if (is_invalid_range(offset, size)) {
+		return -EINVAL;
+	}
+
+	if (!size) {
+		return 0;
+	}
+
+	/**
+	 * Writes occur in three steps:
+	 * A - Chunk of data that needs padding on the left
+	 * B - Chunk of data that is page aligned
+	 * C - Chunk of data that needs padding on the right
+	 * |---------|---------|---------|---------|
+	 *  ~~~~###############################~~~~
+	 *      <-A-> <--------B--------> <-C->
+	 */
+	if (left_pad_len) {
+		/**
+		 * Fill the buffer with left padding and the A part of
+		 * the data.
+		 */
+		write_count = MIN(size, PAGE_SIZE - left_pad_len);
+		memset(page_buffer, ERASE_VALUE, left_pad_len);
+		memcpy(page_buffer + left_pad_len, data, write_count);
+
+		/**
+		 * If the data does not exceed the boundary of the page,
+		 * we also need padding on the right.
+		 */
+		if (left_pad_len + size < PAGE_SIZE) {
+			memset(page_buffer + left_pad_len + size,
+					ERASE_VALUE,
+					PAGE_SIZE - size - left_pad_len);
+		}
+
+		flash_range_program(offset & FLASH_PAGE_MASK, page_buffer,
+					PAGE_SIZE);
+		offset = (offset & FLASH_PAGE_MASK) + PAGE_SIZE;
+		size -= write_count;
+		_data += write_count;
+
+	}
+
+	if (size > PAGE_SIZE) {
+		/**
+		 * The B part of the data is page aligned, so it can be
+		 * written with pico-sdk API.
+		 */
+		flash_range_program(offset, _data, size & FLASH_PAGE_MASK);
+		size -= size & FLASH_PAGE_MASK;
+		offset += size & FLASH_PAGE_MASK;
+		_data += size & FLASH_PAGE_MASK;
+	}
+
+	if (size) {
```

Comment:
might be better to test if size > 0 as well:
MISRA C:2004, 13.2 - Tests of a value against zero should be made explicit, unless the operand is effectively Boolean.